src/function/gds/output_writer.cpp

```diff
@@ -104,99 +107,139 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
-            // more parent edges in the BFS graph AND:
-            // (ii.1) if this is the first element in the stack (curPath.size() == 1), i.e., we
-            // are enumerating the parents of the destination, then we should switch to
-            // forward-tracking if the next parent has visited the destination at a length
-            // that's greater than or equal to the lower bound of the recursive join. Otherwise,
-            // we'll enumerate paths that are smaller than the lower; OR
-            // (ii.2) if this is not the first element in the stack, i.e., then we should switch
-            // to forward tracking only if the next parent of the top node in the stack has the
-            // same iter value as the current parent. That's because the levels/iter need to
-            // decrease by 1 each time we add a new node in the stack.
-            if (top->getNextPtr() != nullptr &&
-                ((curPath.size() == 1 && top->getNextPtr()->getIter() >= info.lowerBound) ||
-                    top->getNextPtr()->getIter() == top->getIter())) {
-                curPath[curPath.size() - 1] = top->getNextPtr();
-                backtracking = false;
+            auto next = getTop(curPath)->getNextPtr();
+            while (true) {
+                if (!canNextReplacePathTop(next, curPath)) {
+                    curPath.pop_back();
+                    break;
+                }
+                // Further check next against path node mask (predicate).
+                if (!checkNewPathElement(curPath, next)) {
+                    next = next->getNextPtr();
+                    continue;
+                }
+                // Next is a valid path element. Push into stack and switch to forward track.
+                curPath[curPath.size() - 1] = next;
                 if (curPath.size() == 1) {
                     setLength(lengthVector.get(), curPath[0]->getIter());
                 }
-            } else {
-                curPath.pop_back();
+                backtracking = false;
+                break;
             }
         } else {
-            auto parent = bfsGraph.getInitialParentAndNextPtr(topNodeID);
-            while (parent->getIter() != top->getIter() - 1) {
+            auto top = getTop(curPath);
+            auto parent = bfsGraph.getInitialParentAndNextPtr(top->getNodeID());
+            while (true) {
+                if (parent == nullptr) {
+                    // No more forward tracking candidates. Switch to backward tracking.
+                    backtracking = true;
+                    break;
+                }
+                if (parent->getIter() == top->getIter() - 1 &&
+                    checkNewPathElement(curPath, parent)) {
+                    // A forward tracking candidate should decrease the iteration by one and also
+                    // pass node predicate checking.
+                    curPath.push_back(parent);
+                    backtracking = false;
```

Comment:
It looks like this backtracking = false line is redundant because are you not guaranteed that backtracking is false here?